File: langs/swift/Sources/VMx/Collections/KeyedServicedObservableCollection.swift
//
// KeyedServicedObservableCollection — ordered, hub-aware collection with a
// captured-key index. See spec/21-collections.md §2.8 and ADR-0097.
//
import Combine
import Foundation

/// Errors reported by keyed collection invariant validation.
public enum KeyedServicedCollectionError: Error {
    /// A candidate mutation would leave two memberships with the same key.
    case duplicateKey
}

/// An ordered serviced collection with expected-O(1) lookup by a captured key.
///
/// Keys are projected before mutation, captured for the lifetime of each
/// membership, and never recomputed by lookup, movement, or removal. Items stay
/// caller-owned. Every effective mutation commits the ordered store and key
/// index before it publishes locally and then to the optional external hub.
public final class KeyedServicedObservableCollection<Key: Hashable, T>: Sequence {
    private let hub: (any MessageHubProtocol)?
    private let keyOf: (T) throws -> Key
    private var items: [T] = []
    private var capturedKeys: [Key] = []
    private var keyToIndex: [Key: Int] = [:]
    private let subject = PassthroughSubject<CollectionChangedMessage<T>, Never>()
    private let typeName = "KeyedServicedObservableCollection"

    /// Create a keyed serviced collection.
    ///
    /// - Parameters:
    ///   - keyOf: A throwing projector whose successful result is captured for
    ///     each new or explicitly replaced membership.
    ///   - hub: An optional external message hub.
    public init(
        keyOf: @escaping (T) throws -> Key,
        hub: (any MessageHubProtocol)? = nil
    ) {
        self.keyOf = keyOf
        self.hub = hub
    }

    /// Hot local publisher for collection changes.
    public var collectionChanged: AnyPublisher<CollectionChangedMessage<T>, Never> {
        subject.eraseToAnyPublisher()
    }

    /// Number of ordered memberships.
    public var count: Int { items.count }

    /// Return the item at `index`, or `nil` when out of bounds.
    public func at(_ index: Int) -> T? {
        guard items.indices.contains(index) else { return nil }
        return items[index]
    }

    /// Return a shallow ordered snapshot.
    public func toArray() -> [T] { items }

    /// Iterate over a snapshot of the current order.
    public func makeIterator() -> IndexingIterator<[T]> {
        items.makeIterator()
    }

    /// Return the item captured under `key`, or `nil` on a miss.
    public func get(_ key: Key) -> T? {
        guard let index = keyToIndex[key] else { return nil }
        return items[index]
    }

    /// Return whether a membership has captured `key`.
    public func containsKey(_ key: Key) -> Bool {
        keyToIndex[key] != nil
    }

    /// Append a new membership after projecting and validating its key.
    public func append(_ item: T) throws {
        let key = try keyOf(item)
        guard keyToIndex[key] == nil else {
            throw KeyedServicedCollectionError.duplicateKey
        }

        let index = items.count
        items.append(item)
        capturedKeys.append(key)
        keyToIndex[key] = index
        emit(.forAdd(sender: self, senderName: typeName, item: item, index: index))
    }

    /// Remove and return the final membership, or `nil` when empty.
    @discardableResult
    public func removeLast() -> T? {
        guard !items.isEmpty else { return nil }
        let index = items.count - 1
        let item = items.removeLast()
        let key = capturedKeys.removeLast()
        keyToIndex.removeValue(forKey: key)
        emit(.forRemove(sender: self, senderName: typeName, item: item, index: index))
        return item
    }

    /// Remove the membership at `index`.
    ///
    /// - Precondition: `index` is a valid collection index.
    public func removeAt(_ index: Int) {
        precondition(
            items.indices.contains(index),
            "KeyedServicedObservableCollection.removeAt: index \(index) out of bounds (count: \(items.count))"
        )
        removeMembership(at: index)
    }

    /// Replace one membership and explicitly recapture its key.
    ///
    /// - Precondition: `index` is a valid collection index.
    public func replace(at index: Int, with newItem: T) throws {
        precondition(
            items.indices.contains(index),
            "KeyedServicedObservableCollection.replace: index \(index) out of bounds (count: \(items.count))"
        )

        let newKey = try keyOf(newItem)
        if let existingIndex = keyToIndex[newKey], existingIndex != index {
            throw KeyedServicedCollectionError.duplicateKey
        }

        let oldItem = items[index]
        let oldKey = capturedKeys[index]
        items[index] = newItem
        capturedKeys[index] = newKey
        // Remove before insert even for equal keys. Hashable reference keys can
        // be distinct equal objects, and explicit replacement recaptures the
        // newly projected key object rather than retaining the retired one.
        keyToIndex.removeValue(forKey: oldKey)
        keyToIndex[newKey] = index
        emit(.forReplace(
            sender: self,
            senderName: typeName,
            newItem: newItem,
            oldItem: oldItem,
            index: index
        ))
    }

    /// Alias for indexed replacement.
    public func setAt(_ index: Int, _ newItem: T) throws {
        try replace(at: index, with: newItem)
    }

    /// Atomically replace all memberships after materialization, projection,
    /// and complete uniqueness validation.
    public func replaceAll<S: Sequence>(_ newItems: S) throws where S.Element == T {
        let newItemSnapshot = Array(newItems)
        var newKeySnapshot: [Key] = []
        newKeySnapshot.reserveCapacity(newItemSnapshot.count)
        var newIndex: [Key: Int] = [:]
        newIndex.reserveCapacity(newItemSnapshot.count)

        for (index, item) in newItemSnapshot.enumerated() {
            let key = try keyOf(item)
            guard newIndex[key] == nil else {
                throw KeyedServicedCollectionError.duplicateKey
            }
            newKeySnapshot.append(key)
            newIndex[key] = index
        }

        guard !items.isEmpty || !newItemSnapshot.isEmpty else { return }
        items = newItemSnapshot
        capturedKeys = newKeySnapshot
        keyToIndex = newIndex
        emit(.forReset(sender: self, senderName: typeName))
    }

    /// Insert a missing projected key or replace its existing membership.
    ///
    /// - Returns: `true` for Add; `false` for in-place Replace.
    @discardableResult
    public func upsert(_ item: T) throws -> Bool {
        let key = try keyOf(item)
        if let index = keyToIndex[key] {
            let oldItem = items[index]
            let oldKey = capturedKeys[index]
            items[index] = item
            capturedKeys[index] = key
            keyToIndex.removeValue(forKey: oldKey)
            keyToIndex[key] = index
            emit(.forReplace(
                sender: self,
                senderName: typeName,
                newItem: item,
                oldItem: oldItem,
                index: index
            ))
            return false
        }

        let index = items.count
        items.append(item)
        capturedKeys.append(key)
        keyToIndex[key] = index
        emit(.forAdd(sender: self, senderName: typeName, item: item, index: index))
        return true
    }

    /// Delete the membership captured under `key`.
    @discardableResult
    public func delete(_ key: Key) -> Bool {
        guard let index = keyToIndex[key] else { return false }
        removeMembership(at: index)
        return true
    }

    /// Move a membership between pre-move positions without reprojecting it.
    public func move(from oldIndex: Int, to newIndex: Int) throws {
        guard items.indices.contains(oldIndex) else {
            throw VMCollectionIndexError(index: oldIndex, count: items.count)
        }
        guard items.indices.contains(newIndex) else {
            throw VMCollectionIndexError(index: newIndex, count: items.count)
        }
        guard oldIndex != newIndex else { return }

        let item = items.remove(at: oldIndex)
        let key = capturedKeys.remove(at: oldIndex)
        items.insert(item, at: newIndex)
        capturedKeys.insert(key, at: newIndex)
        repairIndex(
            from: Swift.min(oldIndex, newIndex),
            through: Swift.max(oldIndex, newIndex)
        )
        emit(.forMove(
            sender: self,
            senderName: typeName,
            item: item,
            from: oldIndex,
            to: newIndex
        ))
    }

    /// Remove every membership. Empty clear is silent.
    public func clear() {
        guard !items.isEmpty else { return }
        items.removeAll()
        capturedKeys.removeAll()
        keyToIndex.removeAll()
        emit(.forReset(sender: self, senderName: typeName))
    }

    private func removeMembership(at index: Int) {
        let item = items.remove(at: index)
        let key = capturedKeys.remove(at: index)
        keyToIndex.removeValue(forKey: key)
        if index < capturedKeys.count {
            repairIndex(from: index, through: capturedKeys.count - 1)
        }
        emit(.forRemove(sender: self, senderName: typeName, item: item, index: index))
    }

    private func repairIndex(from start: Int, through end: Int) {
        guard start <= end else { return }
        for index in start...end {
            keyToIndex[capturedKeys[index]] = index
        }
    }

    private func emit(_ message: CollectionChangedMessage<T>) {
        subject.send(message)
        hub?.send(message)
    }
}

extension KeyedServicedObservableCollection where T: Equatable {
    /// Remove the first equal membership without reprojecting stored items.
    @discardableResult
    public func remove(_ item: T) -> Bool {
        guard let index = items.firstIndex(of: item) else { return false }
        removeMembership(at: index)
        return true
    }
}

// MARK: - Aggregate membership source (reference items only)

extension KeyedServicedObservableCollection: ObservableMembershipSource where T: AnyObject {
    public typealias Item = T

    public func snapshot() -> [T] { toArray() }

    public func subscribeMembership(_ callback: @escaping () -> Void) -> AnyCancellable {
        collectionChanged.sink { _ in callback() }
    }
}

File: langs/typescript/src/collections/keyedServicedObservableCollection.ts
/**
 * Ordered serviced collection with expected O(1) captured-key lookup.
 *
 * Items remain caller-owned. Every mutation settles the ordered items,
 * captured keys, and key-to-index map before local and optional hub delivery.
 *
 * See spec/21-collections.md §§2.8–2.15 and ADR-0097.
 */
import { Observable, Subject } from "rxjs";
import type { IMessageHub } from "../services/messageHub.js";
import { CollectionChangedMessage } from "../messages/collectionChanged.js";

export interface KeyedServicedObservableCollectionOptions<TKey, TItem> {
  readonly keyOf: (item: TItem) => TKey;
  readonly hub?: IMessageHub | null;
}

export class KeyedServicedObservableCollection<TKey, TItem> {
  readonly #keyOf: (item: TItem) => TKey;
  readonly #hub: IMessageHub | null;
  readonly #subject = new Subject<CollectionChangedMessage<TItem>>();
  #items: TItem[] = [];
  #keys: TKey[] = [];
  #indexByKey = new Map<TKey, number>();

  constructor(options: KeyedServicedObservableCollectionOptions<TKey, TItem>) {
    this.#keyOf = options.keyOf;
    this.#hub = options.hub ?? null;
  }

  get collectionChanged(): Observable<CollectionChangedMessage<TItem>> {
    return this.#subject.asObservable();
  }

  get length(): number {
    return this.#items.length;
  }

  at(index: number): TItem | undefined {
    return this.#items[index];
  }

  toArray(): TItem[] {
    return [...this.#items];
  }

  [Symbol.iterator](): IterableIterator<TItem> {
    return this.#items[Symbol.iterator]();
  }

  get(key: TKey): TItem | undefined {
    const index = this.#indexByKey.get(key);
    return index === undefined ? undefined : this.#items[index];
  }

  has(key: TKey): boolean {
    return this.#indexByKey.has(key);
  }

  push(item: TItem): void {
    const key = this.#keyOf(item);
    this.#throwIfDuplicate(key);
    const index = this.#items.length;
    this.#items.push(item);
    this.#keys.push(key);
    this.#indexByKey.set(key, index);
    this.#emit(CollectionChangedMessage.forAdd(this, item, index));
  }

  pop(): TItem | undefined {
    if (this.#items.length === 0) return undefined;
    const index = this.#items.length - 1;
    const item = this.#items[index] as TItem;
    const key = this.#keys[index] as TKey;
    this.#items.pop();
    this.#keys.pop();
    this.#indexByKey.delete(key);
    this.#emit(CollectionChangedMessage.forRemove(this, item, index));
    return item;
  }

  remove(item: TItem): boolean {
    const index = this.#items.indexOf(item);
    if (index === -1) return false;
    this.#removeAt(index);
    return true;
  }

  removeAt(index: number): void {
    this.#validateIndex(index);
    this.#removeAt(index);
  }

  delete(key: TKey): boolean {
    const index = this.#indexByKey.get(key);
    if (index === undefined) return false;
    this.#removeAt(index);
    return true;
  }

  /**
   * Apply native Array.splice normalization to an atomic candidate state.
   * Retained memberships keep their captured keys; inserted items are projected
   * once before commit and may reuse keys removed by the same operation.
   */
  splice(start: number, deleteCount?: number, ...newItems: TItem[]): TItem[] {
    const deleteCountProvided = arguments.length >= 2;
    const insertedKeys = newItems.map((item) => this.#keyOf(item));
    const candidateItems = [...this.#items];
    const candidateKeys = [...this.#keys];
    const resolvedStart = resolveSpliceStart(start, candidateItems.length);
    let removedItems: TItem[];

    if (deleteCountProvided) {
      const effectiveDeleteCount = deleteCount ?? 0;
      removedItems = candidateItems.splice(start, effectiveDeleteCount, ...newItems);
      candidateKeys.splice(start, effectiveDeleteCount, ...insertedKeys);
    } else {
      removedItems = candidateItems.splice(start);
      candidateKeys.splice(start);
    }

    const candidateIndex = this.#buildIndex(candidateKeys);
    if (removedItems.length === 0 && newItems.length === 0) return removedItems;

    this.#items = candidateItems;
    this.#keys = candidateKeys;
    this.#indexByKey = candidateIndex;
    if (removedItems.length === 1 && newItems.length === 0) {
      this.#emit(
        CollectionChangedMessage.forRemove(
          this,
          removedItems[0] as TItem,
          resolvedStart,
        ),
      );
    } else {
      this.#emit(CollectionChangedMessage.forReset(this));
    }
    return removedItems;
  }

  replace(index: number, newItem: TItem): void {
    this.#validateIndex(index);
    const newKey = this.#keyOf(newItem);
    const owner = this.#indexByKey.get(newKey);
    if (owner !== undefined && owner !== index) this.#throwDuplicate();
    this.#replaceAt(index, newItem, newKey);
  }

  setAt(index: number, newItem: TItem): void {
    this.replace(index, newItem);
  }

  replaceAll(items: Iterable<TItem>): void {
    const snapshot = [...items];
    const keys = snapshot.map((item) => this.#keyOf(item));
    const indexByKey = this.#buildIndex(keys);
    if (this.#items.length === 0 && snapshot.length === 0) return;

    this.#items = snapshot;
    this.#keys = keys;
    this.#indexByKey = indexByKey;
    this.#emit(CollectionChangedMessage.forReset(this));
  }

  move(fromIndex: number, toIndex: number): void {
    this.#validateIndex(fromIndex);
    this.#validateIndex(toIndex);
    if (fromIndex === toIndex) return;

    const item = this.#items[fromIndex] as TItem;
    const key = this.#keys[fromIndex] as TKey;
    this.#items.splice(fromIndex, 1);
    this.#keys.splice(fromIndex, 1);
    this.#items.splice(toIndex, 0, item);
    this.#keys.splice(toIndex, 0, key);
    this.#indexByKey = this.#buildIndex(this.#keys);
    this.#emit(
      CollectionChangedMessage.forMove(this, item, fromIndex, toIndex),
    );
  }

  clear(): void {
    if (this.#items.length === 0) return;
    this.#items = [];
    this.#keys = [];
    this.#indexByKey = new Map<TKey, number>();
    this.#emit(CollectionChangedMessage.forReset(this));
  }

  upsert(item: TItem): boolean {
    const key = this.#keyOf(item);
    const index = this.#indexByKey.get(key);
    if (index === undefined) {
      const insertionIndex = this.#items.length;
      this.#items.push(item);
      this.#keys.push(key);
      this.#indexByKey.set(key, insertionIndex);
      this.#emit(CollectionChangedMessage.forAdd(this, item, insertionIndex));
      return true;
    }

    this.#replaceAt(index, item, key);
    return false;
  }

  #removeAt(index: number): void {
    const item = this.#items[index] as TItem;
    this.#items.splice(index, 1);
    this.#keys.splice(index, 1);
    this.#indexByKey = this.#buildIndex(this.#keys);
    this.#emit(CollectionChangedMessage.forRemove(this, item, index));
  }

  #replaceAt(index: number, newItem: TItem, newKey: TKey): void {
    const oldItem = this.#items[index] as TItem;
    const oldKey = this.#keys[index] as TKey;
    this.#items[index] = newItem;
    this.#keys[index] = newKey;
    this.#indexByKey.delete(oldKey);
    this.#indexByKey.set(newKey, index);
    this.#emit(
      CollectionChangedMessage.forReplace(this, newItem, oldItem, index),
    );
  }

  #buildIndex(keys: readonly TKey[]): Map<TKey, number> {
    const result = new Map<TKey, number>();
    keys.forEach((key, index) => {
      if (result.has(key)) this.#throwDuplicate();
      result.set(key, index);
    });
    return result;
  }

  #throwIfDuplicate(key: TKey): void {
    if (this.#indexByKey.has(key)) this.#throwDuplicate();
  }

  #throwDuplicate(): never {
    throw new Error("Duplicate key in KeyedServicedObservableCollection");
  }

  #emit(message: CollectionChangedMessage<TItem>): void {
    this.#subject.next(message);
    this.#hub?.send(message);
  }

  #validateIndex(index: number): void {
    if (
      !Number.isInteger(index) ||
      index < 0 ||
      index >= this.#items.length
    ) {
      throw new RangeError(`Index ${String(index)} out of bounds`);
    }
  }
}

function resolveSpliceStart(start: number, length: number): number {
  const integer = toIntegerOrInfinity(start);
  if (integer < 0) return Math.max(length + integer, 0);
  return Math.min(integer, length);
}

function toIntegerOrInfinity(value: number): number {
  if (Number.isNaN(value) || value === 0) return 0;
  if (!Number.isFinite(value)) return value;
  return Math.trunc(value);
}
